File: prototype_abstract/src/geometry.rs
use euclid::{
    vec2, {Box2D, Point2D, Vector2D},
};

pub struct TileSpace;
pub type TilePosition = Point2D<i32, TileSpace>;
pub type TileVec = Vector2D<i32, TileSpace>;
pub type BoundingBox = Box2D<i32, TileSpace>;

pub fn pos(x: i32, y: i32) -> TilePosition {
    TilePosition::new(x, y)
}

/// South is +y
/// East is +x
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum Direction {
    North = 0,
    East = 1,
    South = 2,
    West = 3,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Axis {
    X,
    Y,
}

impl Direction {
    pub const fn to_vector(self) -> TileVec {
        match self {
            Direction::North => vec2(0, -1),
            Direction::East => vec2(1, 0),
            Direction::South => vec2(0, 1),
            Direction::West => vec2(-1, 0),
        }
    }

    pub const fn opposite(self) -> Direction {
        match self {
            Direction::North => Direction::South,
            Direction::East => Direction::West,
            Direction::South => Direction::North,
            Direction::West => Direction::East,
        }
    }

    pub const fn rotate_cw(self) -> Direction {
        match self {
            Direction::North => Direction::East,
            Direction::East => Direction::South,
            Direction::South => Direction::West,
            Direction::West => Direction::North,
        }
    }

    pub const fn rotate_ccw(self) -> Direction {
        match self {
            Direction::North => Direction::West,
            Direction::East => Direction::North,
            Direction::South => Direction::East,
            Direction::West => Direction::South,
        }
    }

    pub const fn axis(self) -> Axis {
        match self {
            Direction::North | Direction::South => Axis::Y,
            Direction::East | Direction::West => Axis::X,
        }
    }

    pub const fn from_ordinal(ordinal: u8) -> Option<Direction> {
        match ordinal {
            0 => Some(Direction::North),
            1 => Some(Direction::East),
            2 => Some(Direction::South),
            3 => Some(Direction::West),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Copy)]
pub struct Ray {
    pub start_position: TilePosition,
    pub direction: Direction,
}

impl Ray {
    pub const fn new(position: TilePosition, direction: Direction) -> Self {
        Self {
            start_position: position,
            direction,
        }
    }

    pub fn ray_position(&self, position: TilePosition) -> i32 {
        let offset = position - self.start_position;
        let dir_vec = self.direction.to_vector();
        offset.dot(dir_vec)
    }
    /// Get position at index along ray using euclid vector arithmetic
    pub fn get_position(&self, index: i32) -> TilePosition {
        self.start_position + self.direction.to_vector() * index
    }
    pub fn snap(&self, position: TilePosition) -> TilePosition {
        self.get_position(self.ray_position(position))
    }

    pub fn relative_direction(&self, position: TilePosition) -> Option<Direction> {
        let offset = position - self.start_position;
        match self.direction {
            Direction::North | Direction::South => {
                if offset.x == 0 {
                    None
                } else if offset.x > 0 {
                    Some(Direction::East)
                } else {
                    Some(Direction::West)
                }
            }
            Direction::East | Direction::West => {
                if offset.y == 0 {
                    None
                } else if offset.y > 0 {
                    Some(Direction::South)
                } else {
                    Some(Direction::North)
                }
            }
        }
    }
}

pub fn bounds_new(top_left: TilePosition, bottom_right: TilePosition) -> BoundingBox {
    BoundingBox::new(top_left, bottom_right)
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Transform {
    pub flip_x: bool,
    pub flip_y: bool,
    pub swap_x_y: bool,
}

impl Transform {
    pub fn new(flip_x: bool, flip_y: bool, swap_x_y: bool) -> Self {
        Self {
            flip_x,
            flip_y,
            swap_x_y,
        }
    }

    pub fn identity() -> Self {
        Default::default()
    }

    pub fn transform_position(&self, position: TilePosition) -> TilePosition {
        let mut result = position;

        if self.swap_x_y {
            result = pos(result.y, result.x);
        }

        if self.flip_x {
            result = pos(-result.x, result.y);
        }

        if self.flip_y {
            result = pos(result.x, -result.y);
        }

        result
    }

    pub fn transform_direction(&self, dir: Direction) -> Direction {
        let mut ordinal = dir as u8;

        if self.swap_x_y {
            // Swapping x and y for direction vectors:
            // North(0,-1) → (-1,0)=West, East(1,0) → (0,1)=South,
            // South(0,1) → (1,0)=East, West(-1,0) → (0,-1)=North
            ordinal = match ordinal {
                0 => 3,
                1 => 2,
                2 => 1,
                3 => 0,
                _ => ordinal,
            };
        }

        if self.flip_x {
            ordinal = match ordinal {
                1 => 3,
                3 => 1,
                _ => ordinal,
            };
        }

        if self.flip_y {
            ordinal = match ordinal {
                0 => 2,
                2 => 0,
                _ => ordinal,
            };
        }

        Direction::from_ordinal(ordinal).unwrap()
    }

    pub fn all_unique_transforms() -> Vec<Transform> {
        [
            Transform::new(false, false, false), // identity
            Transform::new(true, false, true),   // 90° CW
            Transform::new(true, true, false),   // 180°
            Transform::new(false, true, true),   // 90° CCW
            Transform::new(true, false, false),  // flip X
            Transform::new(true, true, true),    // flip X then 90° CW
            Transform::new(false, true, false),  // flip Y
            Transform::new(false, false, true),  // flip Y then 90° CW (swap only)
        ]
        .to_vec()
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_direction_opposite() {
        assert_eq!(Direction::North.opposite(), Direction::South);
        assert_eq!(Direction::East.opposite(), Direction::West);
        assert_eq!(Direction::South.opposite(), Direction::North);
        assert_eq!(Direction::West.opposite(), Direction::East);
    }

    #[test]
    fn test_direction_ordinal() {
        assert_eq!(Direction::from_ordinal(0), Some(Direction::North));
        assert_eq!(Direction::from_ordinal(1), Some(Direction::East));
        assert_eq!(Direction::from_ordinal(2), Some(Direction::South));
        assert_eq!(Direction::from_ordinal(3), Some(Direction::West));
        assert_eq!(Direction::from_ordinal(4), None);
    }

    #[test]
    fn test_ray_distance() {
        let ray_north = Ray::new(pos(0, 0), Direction::North);
        assert_eq!(ray_north.ray_position(pos(0, -5)), 5);
        assert_eq!(ray_north.ray_position(pos(0, 5)), -5);

        let ray_east = Ray::new(pos(0, 0), Direction::East);
        assert_eq!(ray_east.ray_position(pos(5, 0)), 5);
        assert_eq!(ray_east.ray_position(pos(-5, 0)), -5);

        let ray_south = Ray::new(pos(0, 0), Direction::South);
        assert_eq!(ray_south.ray_position(pos(0, 5)), 5);
        assert_eq!(ray_south.ray_position(pos(0, -5)), -5);

        let ray_west = Ray::new(pos(0, 0), Direction::West);
        assert_eq!(ray_west.ray_position(pos(5, 0)), -5);
        assert_eq!(ray_west.ray_position(pos(-5, 0)), 5);
    }

    #[test]
    fn test_position_at() {
        let ray_north = Ray::new(pos(1, 1), Direction::North);
        assert_eq!(ray_north.get_position(5), pos(1, -4));

        let ray_east = Ray::new(pos(1, 1), Direction::East);
        assert_eq!(ray_east.get_position(5), pos(6, 1));

        let ray_south = Ray::new(pos(1, 1), Direction::South);
        assert_eq!(ray_south.get_position(5), pos(1, 6));

        let ray_west = Ray::new(pos(1, 1), Direction::West);
        assert_eq!(ray_west.get_position(5), pos(-4, 1));
    }

    #[test]
    fn test_snap() {
        let ray_north = Ray::new(pos(1, 1), Direction::North);
        assert_eq!(ray_north.snap(pos(5, -4)), pos(1, -4));

        let ray_east = Ray::new(pos(1, 1), Direction::East);
        assert_eq!(ray_east.snap(pos(6, 5)), pos(6, 1));

        let ray_south = Ray::new(pos(1, 1), Direction::South);
        assert_eq!(ray_south.snap(pos(5, 6)), pos(1, 6));

        let ray_west = Ray::new(pos(1, 1), Direction::West);
        assert_eq!(ray_west.snap(pos(-4, 5)), pos(-4, 1));
    }

    #[test]
    fn test_transform_position() {
        let pos1 = pos(2, 3);

        assert_eq!(Transform::identity().transform_position(pos1), pos(2, 3));

        assert_eq!(
            Transform::new(true, false, false).transform_position(pos1),
            pos(-2, 3)
        );
        assert_eq!(
            Transform::new(false, true, false).transform_position(pos1),
            pos(2, -3)
        );
        assert_eq!(
            Transform::new(true, true, false).transform_position(pos1),
            pos(-2, -3)
        );

        assert_eq!(
            Transform::new(true, false, true).transform_position(pos1),
            pos(-3, 2)
        );
        assert_eq!(
            Transform::new(true, true, false).transform_position(pos1),
            pos(-2, -3)
        );
        assert_eq!(
            Transform::new(false, true, true).transform_position(pos1),
            pos(3, -2)
        );
    }

    #[test]
    fn test_transform_direction() {
        assert_eq!(
            Transform::identity().transform_direction(Direction::North),
            Direction::North
        );
        assert_eq!(
            Transform::identity().transform_direction(Direction::East),
            Direction::East
        );

        assert_eq!(
            Transform::new(true, false, false).transform_direction(Direction::East),
            Direction::West
        );
        assert_eq!(
            Transform::new(true, false, false).transform_direction(Direction::West),
            Direction::East
        );
        assert_eq!(
            Transform::new(true, false, false).transform_direction(Direction::North),
            Direction::North
        );

        assert_eq!(
            Transform::new(false, true, false).transform_direction(Direction::North),
            Direction::South
        );
        assert_eq!(
            Transform::new(false, true, false).transform_direction(Direction::South),
            Direction::North
        );
        assert_eq!(
            Transform::new(false, true, false).transform_direction(Direction::East),
            Direction::East
        );

        assert_eq!(
            Transform::new(true, false, true).transform_direction(Direction::North),
            Direction::East
        );
        assert_eq!(
            Transform::new(true, false, true).transform_direction(Direction::East),
            Direction::South
        );
        assert_eq!(
            Transform::new(true, true, false).transform_direction(Direction::North),
            Direction::South
        );
        assert_eq!(
            Transform::new(false, true, true).transform_direction(Direction::North),
            Direction::West
        );
    }

    #[test]
    fn test_transform_entity() {
        use crate::{Belt, YELLOW_BELT};

        let belt = Belt::new(Direction::North, YELLOW_BELT);
        let transform = Transform::new(true, false, true);

        let transformed = transform.transform_entity(belt.as_ref());
        let transformed_belt = transformed.as_belt().unwrap();

        assert_eq!(transformed_belt.direction, Direction::East);
        assert_eq!(transformed_belt.tier, YELLOW_BELT);
    }

    #[test]
    fn test_combined_flip_rotation() {
        let pos_test = pos(3, 4);
        // Horizontal flip then Right rotation: (x,y) → (-x,y) → (-y,-x)
        let transform = Transform::new(true, true, true);
        let result = transform.transform_position(pos_test);

        assert_eq!(result, pos(-4, -3));

        let dir_test = Direction::North;
        let dir_result = transform.transform_direction(dir_test);
        assert_eq!(dir_result, Direction::East);
    }
}

File: prototype_mod/common/geometry.ts
export interface Point2D<T> {
  readonly x: number
  readonly y: number
}

export interface Vector2D<T> {
  readonly x: number
  readonly y: number
}

export interface TilePosition {
  readonly x: number
  readonly y: number
}
export interface TileVec {
  readonly x: number
  readonly y: number
}

export function pos(x: number, y: number): TilePosition {
  return { x, y }
}

export function vec2(x: number, y: number): TileVec {
  return { x, y }
}

export function addVec(p: TilePosition, v: TileVec): TilePosition {
  return { x: p.x + v.x, y: p.y + v.y }
}

export function subPos(p1: TilePosition, p2: TilePosition): TileVec {
  return { x: p1.x - p2.x, y: p1.y - p2.y }
}

export function mulVec(v: TileVec, scalar: number): TileVec {
  return { x: v.x * scalar, y: v.y * scalar }
}

export function dotVec(v1: TileVec, v2: TileVec): number {
  return v1.x * v2.x + v1.y * v2.y
}

export function posEquals(p1: TilePosition, p2: TilePosition): boolean {
  return p1.x === p2.x && p1.y === p2.y
}

export const enum Direction {
  North = 0,
  East = 1,
  South = 2,
  West = 3,
}

export const enum Axis {
  X = 0,
  Y = 1,
}

export function directionToVector(dir: Direction): TileVec {
  switch (dir) {
    case Direction.North:
      return vec2(0, -1)
    case Direction.East:
      return vec2(1, 0)
    case Direction.South:
      return vec2(0, 1)
    case Direction.West:
      return vec2(-1, 0)
  }
}

export function oppositeDirection(dir: Direction): Direction {
  switch (dir) {
    case Direction.North:
      return Direction.South
    case Direction.East:
      return Direction.West
    case Direction.South:
      return Direction.North
    case Direction.West:
      return Direction.East
  }
}

export function directionAxis(dir: Direction): Axis {
  switch (dir) {
    case Direction.North:
    case Direction.South:
      return Axis.Y
    case Direction.East:
    case Direction.West:
      return Axis.X
  }
}

export function axisSign(dir: Direction): number {
  switch (dir) {
    case Direction.North:
      return -1
    case Direction.South:
      return 1
    case Direction.East:
      return 1
    case Direction.West:
      return -1
  }
}

export interface Ray {
  readonly crossAxisValue: number
  readonly direction: Direction
}

export function createRay(position: TilePosition, direction: Direction): Ray {
  const crossAxisValue =
    directionAxis(direction) === Axis.Y ? position.x : position.y
  return { crossAxisValue, direction }
}

export function rayDirection(ray: Ray): Direction {
  return ray.direction
}

export function rayPosition(ray: Ray, position: TilePosition): number {
  return directionAxis(ray.direction) === Axis.Y ? position.y : position.x
}

export function getRayPosition(ray: Ray, pos: number): TilePosition {
  return directionAxis(ray.direction) === Axis.Y
    ? { x: ray.crossAxisValue, y: pos }
    : { x: pos, y: ray.crossAxisValue }
}

export function raySnap(ray: Ray, position: TilePosition): TilePosition {
  return getRayPosition(ray, rayPosition(ray, position))
}

export function isBeforeOnRay(ray: Ray, a: number, b: number): boolean {
  return (b - a) * axisSign(ray.direction) > 0
}

export function rayRelativeDirection(
  ray: Ray,
  position: TilePosition,
): Direction | undefined {
  const cross =
    directionAxis(ray.direction) === Axis.Y
      ? position.x - ray.crossAxisValue
      : position.y - ray.crossAxisValue

  switch (ray.direction) {
    case Direction.North:
    case Direction.South:
      if (cross === 0) return undefined
      return cross > 0 ? Direction.East : Direction.West
    case Direction.East:
    case Direction.West:
      if (cross === 0) return undefined
      return cross > 0 ? Direction.South : Direction.North
  }
}

export interface BoundingBox {
  readonly left_top: TilePosition
  readonly right_bottom: TilePosition
}

export function boundsNew(
  topLeft: TilePosition,
  bottomRight: TilePosition,
): BoundingBox {
  return { left_top: topLeft, right_bottom: bottomRight }
}

export function boundsUnion(b1: BoundingBox, b2: BoundingBox): BoundingBox {
  return {
    left_top: pos(
      Math.min(b1.left_top.x, b2.left_top.x),
      Math.min(b1.left_top.y, b2.left_top.y),
    ),
    right_bottom: pos(
      Math.max(b1.right_bottom.x, b2.right_bottom.x),
      Math.max(b1.right_bottom.y, b2.right_bottom.y),
    ),
  }
}

export function boundsContains(
  bounds: BoundingBox,
  position: TilePosition,
): boolean {
  return (
    position.x >= bounds.left_top.x &&
    position.x <= bounds.right_bottom.x &&
    position.y >= bounds.left_top.y &&
    position.y <= bounds.right_bottom.y
  )
}
